components/ipfs/ipfs_utils.cc

```diff
@@ -460,47 +464,110 @@ std::string GetRegistryDomainFromIPNS(const GURL& url) {
+      buff.push_back('.');
+    } else if (!is_last && chars[i] == '-' && chars[i + 1] == '-') {
+      buff.push_back('-');
+      i++;
+    } else {
+      buff.push_back(chars[i]);
+    }
   }
+  std::string result = std::string(buff.begin(), buff.end());
+  return result;
+}
 
+absl::optional<GURL> ExtractSourceFromGatewayHost(const GURL& url) {
   std::vector<std::string> host_parts = base::SplitStringUsingSubstr(
       url.host(), ".", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
 
-  if (host_parts.size() > 2 && IsValidCID(host_parts.at(0)) &&
-      host_parts.at(1) == "ipfs") {
-    GURL final_url = GURL("ipfs://" + host_parts.at(0) + url.path());
+  if (host_parts.size() > 2) {
+    GURL final_url;
+    if (host_parts.at(1) == "ipfs" && IsValidCID(host_parts.at(0))) {
+      final_url = GURL("ipfs://" + host_parts.at(0) + url.path());
+    } else if (host_parts.at(1) == "ipns" && IsValidIPNSCID(host_parts.at(0))) {
+      final_url = GURL("ipns://" + host_parts.at(0) + url.path());
+    } else if (host_parts.at(1) == "ipns") {
+      std::string decoded = DecodeSingleLabelForm(host_parts.at(0));
+      final_url = GURL("https://" + decoded + url.path());
+    }
+
+    if (!final_url.is_valid()) {
+      return absl::nullopt;
+    }
+
     GURL::Replacements replacements;
     replacements.SetQueryStr(url.query_piece());
     replacements.SetRefStr(url.ref_piece());
     return final_url.ReplaceComponents(replacements);
   }
+  return absl::nullopt;
+}
 
+absl::optional<GURL> ExtractSourceFromGatewayPath(const GURL& url) {
   std::vector<std::string> path_parts = base::SplitStringUsingSubstr(
       url.path(), "/", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
 
-  if (path_parts.size() >= 2 && path_parts.at(0) == "ipfs" &&
-      IsValidCID(path_parts.at(1))) {
+  if (path_parts.size() >= 2) {
+    GURL final_url;
+
     std::string final_path;
     if (path_parts.size() >= 3) {
       std::vector<std::string> final_path_parts(path_parts.begin() + 2,
                                                 path_parts.end());
       final_path = "/" + base::JoinString(final_path_parts, "/");
     }
 
-    GURL final_url = GURL("ipfs://" + path_parts.at(1) + final_path);
+    if (path_parts.at(0) == "ipfs" && IsValidCID(path_parts.at(1))) {
+      final_url = GURL("ipfs://" + path_parts.at(1) + final_path);
+    } else if (path_parts.at(0) == "ipns" && IsValidIPNSCID(path_parts.at(1))) {
+      final_url = GURL("ipns://" + path_parts.at(1) + final_path);
+    } else if (path_parts.at(0) == "ipns") {
+      std::string decoded = DecodeSingleLabelForm(path_parts.at(1));
+      final_url = GURL("https://" + decoded + final_path);
+    }
+
+    if (!final_url.is_valid()) {
+      return absl::nullopt;
+    }
+
     GURL::Replacements replacements;
     replacements.SetQueryStr(url.query_piece());
     replacements.SetRefStr(url.ref_piece());
     return final_url.ReplaceComponents(replacements);
   }
-
   return absl::nullopt;
 }
```

Comment:
ExtractSourceFromGatewayHost and ExtractSourceFromGatewayPath look very similar, can it be refactored to smth like ExtractSource(scheme, host, path) ?

Reply:
This was the first variant. I've separated it to make overall code more clear, because it was looking too complicated.